src/main/kotlin/com/openlattice/organizations/HazelcastOrganizationService.kt

```diff
@@ -285,44 +290,47 @@ class HazelcastOrganizationService(
         }
     }
 
+    /**
+     * Do not invoke this directly as it bypasses checks in [HazelcastOrganizationService.addMembers]
+     *
+     */
     private fun addMembers(
             orgAclKey: AclKey,
-            membersToAdd: Set<Principal>,
+            membersToAdd: Map<Principal, AclKey>,
             profiles: Map<Principal, Map<String, Set<String>>>
     ): Set<Principal> {
         require(orgAclKey.size == 1) { "Organization acl key should only be of length 1" }
-
-        val members = if (membersToAdd is Serializable) membersToAdd else membersToAdd.toSet()
-
-        val nonUserPrincipals = members.filter { it.type != PrincipalType.USER }
-        require(nonUserPrincipals.isEmpty()) { "Cannot add non-users principals $nonUserPrincipals to organization $orgAclKey" }
+        val members = membersToAdd.keys.toSet()
         val organizationId = orgAclKey[0]
-
-
         val newMembers: Set<Principal> = organizations.executeOnKey(organizationId, OrganizationEntryProcessor {
-            val newMembers = members.filter { member -> !it.members.contains(member) }.toSet()
+            val newMembers = members - it.members
             it.members.addAll(newMembers)
-            return@OrganizationEntryProcessor Result(newMembers, newMembers.isNotEmpty())
+            //We're always persisting here.
+            return@OrganizationEntryProcessor Result(newMembers)
         }) as Set<Principal>
 
-        if (newMembers.isNotEmpty()) {
-            grantOrganizationPrincipals(organizationId, newMembers, orgAclKey, profiles)
-        }
+        //Always trigger as this won't cause a write to organizations table.
+        grantOrganizationPrincipals(
+                organizationId,
+                membersToAdd,
```

Comment:
Why not `newMembers`?

Reply:
To be fail safe. There was a weird bug where if grantOrganizationPrincipals fails (say stack reboot) you can end up with a member of the org that isn't visible as a member (listing members uses grant). Adding the user again will cause the grant to be re-attempted even if members list is not modified.